libfreerdp/core/proxy.c

```diff
@@ -255,3 +284,181 @@ BOOL http_proxy_connect(BIO* bufferedBio, const char* hostname, UINT16 port)
 	return TRUE;
 }
 
+static int recv_socks_reply(BIO* bufferedBio, BYTE* buf, int len, char* reason, BYTE checkVer)
+{
+	int status;
+
+	for(;;)
+	{
+		status = BIO_read(bufferedBio, buf, len);
+		if (status > 0)
+			break;
+
+		if (status < 0)
+		{
+			/* Error? */
+			if (BIO_should_retry(bufferedBio))
+			{
+				USleep(100);
+				continue;
+			}
+
+			WLog_ERR(TAG, "Failed reading %s reply from SOCKS proxy (Status %d)", reason, status);
+			return -1;
+		}
+
+		if (status == 0)
+		{
+			/* Error? */
+			WLog_ERR(TAG, "Failed reading %s reply from SOCKS proxy (BIO_read returned zero)", reason);
+			return -1;
+		}
+	}
+	
+	if (status < 2)
+	{
+		WLog_ERR(TAG, "SOCKS Proxy reply packet too short (%s)", reason);
+		return -1;
+	}
+
+	if (buf[0] != checkVer)
+	{
+		WLog_ERR(TAG, "SOCKS Proxy version is not 5 (%s)", reason);
+		return -1;
+	}
+
+	return status;
+}
+
+/* CONN REQ replies in enum. order */
+static const char *rplstat[] =
```

Comment:
should be at start of file

Reply:
Well I kept it here because it's SOCKS only (as the file treat http and SOCKS proxy), but I can move it